File: app/src/main/java/sixue/naivereader/NetProviderManagerActivity.java
package sixue.naivereader;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.ListView;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import sixue.naivereader.provider.NetProvider;
import sixue.naivereader.provider.NetProviderCollections;

public class NetProviderManagerActivity extends AppCompatActivity {

    private List<NetProvider> netProviders;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_net_provider_manager);

        netProviders = new ArrayList<>(NetProviderCollections.getProviders(this));

        ListView listView = findViewById(R.id.list_net_providers);
        listView.setAdapter(new BaseAdapter() {
            @Override
            public int getCount() {
                return netProviders.size();
            }

            @Override
            public Object getItem(int i) {
                return null;
            }

            @Override
            public long getItemId(int i) {
                return 0;
            }

            @Override
            public View getView(int i, View view, ViewGroup viewGroup) {
                if (view == null)
                    view = LayoutInflater.from(NetProviderManagerActivity.this).inflate(R.layout.listviewitem_provider, viewGroup, false);

                final TextView name = view.findViewById(R.id.name);
                final Switch sw = view.findViewById(R.id.sw);

                final NetProvider netProvider = netProviders.get(i);
                name.setText(netProvider.getProviderName());
                sw.setChecked(netProvider.isActive());
                sw.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        netProvider.setActive(sw.isChecked());
                        NetProviderCollections.saveSettings(NetProviderManagerActivity.this);
                    }
                });
                sw.setChecked(netProvider.isActive());

                return view;
            }
        });

        Button cg = findViewById(R.id.garbage_button);
        cg.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                BookLoader.getInstance().clearGarbage();
                Toast.makeText(NetProviderManagerActivity.this, "[GARBAGE] clear.", Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/sixue/naivereader/NetProviderManagerActivity.kt
package sixue.naivereader

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SwitchCompat
import sixue.naivereader.provider.NetProvider
import sixue.naivereader.provider.NetProviderCollections.getProviders
import sixue.naivereader.provider.NetProviderCollections.saveSettings
import java.util.*

class NetProviderManagerActivity : AppCompatActivity() {
    private lateinit var netProviders: List<NetProvider>
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_net_provider_manager)
        netProviders = ArrayList(getProviders(this))
        val listView = findViewById<ListView>(R.id.list_net_providers)
        listView.adapter = object : BaseAdapter() {
            override fun getCount(): Int {
                return netProviders.size
            }

            override fun getItem(i: Int): Any {
                return ""
            }

            override fun getItemId(i: Int): Long {
                return 0
            }

            override fun getView(i: Int, convertView: View?, viewGroup: ViewGroup): View {
                val view : View?
                if (convertView == null){
                    view = LayoutInflater.from(this@NetProviderManagerActivity).inflate(R.layout.listviewitem_provider, viewGroup, false)
                    val name = view.findViewById<TextView>(R.id.name)
                    val sw = view.findViewById<SwitchCompat>(R.id.sw)
                    view.tag = ViewHolder(name, sw)
                } else {
                    view = convertView
                }
                val viewHolder: ViewHolder = view!!.tag as ViewHolder
                val netProvider = netProviders[i]
                viewHolder.name.text = netProvider.providerName
                viewHolder.sw.isChecked = netProvider.isActive
                viewHolder.sw.setOnClickListener {
                    netProvider.isActive = viewHolder.sw.isChecked
                    saveSettings(this@NetProviderManagerActivity)
                }
                viewHolder.sw.isChecked = netProvider.isActive
                return view
            }

            inner class ViewHolder(val name: TextView, val sw: SwitchCompat)
        }
        val cg = findViewById<Button>(R.id.garbage_button)
        cg.setOnClickListener {
            val n = BookLoader.clearGarbage {
                Toast.makeText(this, "[GARBAGE] $it", Toast.LENGTH_SHORT).show()
            }
            Toast.makeText(this@NetProviderManagerActivity, "[GARBAGE] Clear <$n>.", Toast.LENGTH_SHORT)
                .show()
        }
    }
}